### ISSUE TITLE
Convert to tflite will change output order for models with multiple outputs

### ISSUE THREAD

**Issue by shayantabatabaee** [tensorflow/tensorflow#33303]
Hello everyone, 

Currently I have a model with one input and multiple outputs. I train my model with tensorflow eager execution mode and saved trained model into keras format. after that I converted model to tflite format to use in mobile devices. The issue is when I converted my model to tflite the output order will changed. I tested same image with keras model and tflite model and the output order is different.
 
**System information**
- OS Platform and Distribution (Mac OS Mojave)
- TensorFlow version : 2.0.0
- Number of outputs: 16
- Base Model: MobilenetV1(0.25)

Here is my outputs for both keras and tflite :

tflite :6157037319861035

keras :6219861085157097

and here is a picture of output layer of my model :

![Screen Shot 2019-10-13 at 3 06 29 PM](https://user-images.githubusercontent.com/28183901/66715108-1c92ba00-edcc-11e9-8f26-eaa147e49e76.png)

**Comment by gadagashwini** [comment]
@shayantabatabaee, Will it be possible to provide the keras model. And also, code snippet used to convert keras to tflite. Thanks!

**Comment by shayantabatabaee** [comment]
Yes , Here is my model : 

```
 base_model = MobileNet(input_shape=(64, 640, 3),
                           alpha=0.5,
                           depth_multiplier=1,
                           include_top=False,
                           weights='imagenet',
                           input_tensor=None,
                           pooling=None)

    x = base_model.output
    x = layers.GlobalAveragePooling2D()(x)
    x = layers.Reshape((1, 1, 512), name='reshape_1')(x)

    digits = []
    for i in range(16):
        temp = layers.Dropout(1e-1, name=str(i) + "_dropout")(x)
        temp = layers.Conv2D(10, 1, 1, padding='same', name=str(i) + '_digits')(temp)
        temp = layers.Reshape((10,), name=str(i) + '_reshape')(temp)
        temp = layers.Activation('softmax', name=str(i) + '_softmax')(temp)
        digits.append(temp)

    model = Model(inputs=base_model.inputs, outputs=digits)

    model.compile(optimizer=optimizers.Adam(), loss='categorical_crossentropy',
                  metrics=['accuracy'])

    model.summary()
    return model

```

and I convert my model using this script:

```
def convert_to_lite_model(save_path, model, tflite_name):
    print("Start converting to lite model")
    converter = tf.lite.TFLiteConverter.from_keras_model(model)
    converter.optimizations = [tf.lite.Optimize.DEFAULT]
    converter.representative_dataset = representative_dataset_gen
    tflite_model = converter.convert()
    open(os.path.join(save_path, tflite_name), "wb").write(tflite_model)

```

**Comment by shayantabatabaee** [comment]
I made some changes in my model, I stack the outputs layer into one layer using tf.stack : 
`outputs=tf.stack(digits, axis=1)` 
It seems that now outputs order is same with my keras model but I have another issue.
this model will be used for OCR of credit card numbers, The accuracy of my keras model is about 98% and everything works fine , but although the outputs is constant now, my model finds the numbers but the orders is not correct. Let's see this example,
I have a credit card with this number : 5022 2**9**1**0** 3**62**0 4580 
my model will find numbers  : 5022 2**2**1**6** 3**09**0 4580.
The numbers is correct with high precision but the orders is not correct and it is happened only for middle numbers.

**Comment by jvishnuvardhan** [comment]
@shayantabatabaee Can you provide 'representative_dataset_gen`? I tried making a [gist](https://colab.sandbox.google.com/gist/jvishnuvardhan/1501dc688045f353b3f52bce8e44b686/untitled640.ipynb) with your code so that we can reproduce the issue. However, 'representative_dataset_gen` is missing. If you cannot provide that dataset, can you please create a simple standalone code to reproduce the issue. Thanks!

**Comment by shayantabatabaee** [comment]
@jvishnuvardhan actually you do not need dataset to reproduce bug, just comment representative_dataset_gen line and you will see the bug, I tested your gist and saw the bug. I use Netron application to see that the orders of outputs layers is not the same with the code. I seemed that second output of model is that 10th output of code, I attached the image. 
<img width="400" alt="Screen Shot 2019-11-15 at 10 41 05 AM" src="https://user-images.githubusercontent.com/28183901/68923965-e6fb3b00-0794-11ea-8d1a-b10d5cfdc5fa.png">
And for the second issue that I saw I just change the aspect ratio of input image , I doubled the height and change the image size to 100 x 500 so the second issue is gone.

**Comment by marno1d** [comment]
I have noticed that the orders of both the inputs and outputs are mixed up after tflite conversion. 

Another problem this creates is when you are feeding in the representative dataset you don't know what order the inputs are in.

**Comment by Cospel** [comment]
This is also partially related to #32180

**Comment by Thesane** [comment]
providing the output_arrays (and input_arrays if needed) will solve the problem as you can guarantee the order by doing so
[https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/lite/TFLiteConverter#from_keras_model_file](https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/lite/TFLiteConverter#from_keras_model_file)

**Comment by mtngld** [comment]
I am also seeing this with `from_saved_model`, also seems that `output_arrays` solution is no longer supported in TF2

**Comment by Robird** [comment]
i'm facing exactly the same issue.
[lite_chaos_order.zip](https://github.com/tensorflow/tensorflow/files/6014492/lite_chaos_order.zip)
just feed with same random to both h5 and lite model, you can regen the issue（same output value collection but different order）

**Comment by francotheengineer** [comment]
I can reproduce this issue in TF 2.4.1 . Inspecting the list of output tensors from:
 `interpreter = tf.lite.Interpreter(
    model_path='/path/to/tflite/model'
interpreter.get_output_details()` 
Yeilds a out-of-order list, compared to the `.predict` call in tf.keras, specifically, the list has been mirrored. For 4 output tensors: 
tf.keras.Model.predict `output[0] ` is now `output[3] ` in tf.lite
tf.keras.Model.predict `output[1] ` is now `output[2] ` in tf.lite
tf.keras.Model.predict `output[2] ` is now `output[1] ` in tf.lite
tf.keras.Model.predict `output[3] ` is now `output[0] ` in tf.lite
I have not attempted to fix the export to TFlite, since I can deal with this simply by parsing the interpreter output in reverse order.

**Comment by DocGarbanzo** [comment]
I'm seeing the same thing in TF 2.2.0: output 1, 2, 3 is returned as 1, 3, 2.

**Comment by francotheengineer** [comment]
> I am also seeing this with `from_saved_model`, also seems that `output_arrays` solution is no longer supported in TF2
This is correct. 

I have been working for 2 days trying to find a solution to this issue, and have yet to find one. The issues for me is, if I push models to production, and each time the export has a different output order, it can't be used. I'd be keen to contribute to fixing this but will be a while before I can get onto that. Thanks all!

**Comment by johanahlkvist** [comment]
Naming the outputs in keras such that their names' alphabetical order correspond to their desired order in tflite worked well for me.

**Comment by francotheengineer** [comment]
> Naming the outputs in keras such that their names' alphabetical order correspond to their desired order in tflite worked well for me.

Thanks very much for the suggestion! Didn't manage to get this working because the outputs are still out-of-order. If i understand correctly you mean labeling the outputs as such:

`model = tf.keras.Model(inputs=in_tensor, outputs={'1': x1,'2':x2,'3': x3,'4': x4}, name= 'my_model'`

**Comment by johanahlkvist** [comment]
Yes, that's what I mean. Perhaps it is equivalent, but specifically, I did:

```python
# (...)
y1 = Dense(1, name='A')(some_input1)
y2 = Dense(1, name='B')(some_input2)
return Model(x, [y1, y2])
```

**Comment by yujanshrestha** [comment]
Please fix this. I was hoping to use TFLite as a drop in replacement for the TF runtime. However, the unpredictable ordering for multi-output models is surprising behavior and took me a long time to figure out what was going on.

**Comment by sushreebarsa** [comment]
Was able to reproduce the issue in TF v2.5,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/bde60d31524f87e0d1466a4a1d1228b4/untitled21.ipynb#scrollTo=jFMmYHFP_KbA),Thanks !

**Comment by charleswg** [comment]
I managed to get the order back by sorting the index of each output using tflite_interpreter.get_output_details()

**Comment by ydshieh** [comment]
When I use `interpreter.get_output_details()`, I got a list like the following
```
  [
    {'name': 'StatefulPartitionedCall:5', 'index': 714, ...}, 
    {'name': 'StatefulPartitionedCall:4', 'index': 716, ...}, 
    {'name': 'StatefulPartitionedCall:3', 'index': 717, ...}, 
    {'name': 'StatefulPartitionedCall:2', 'index': 734, ...}, 
    {'name': 'StatefulPartitionedCall:1', 'index': 736, ...}, 
    {'name': 'StatefulPartitionedCall:0', 'index': 737, ...}
  ]
```

However, the order of my keras model outputs is in the reversed order, i.e. I need to use number `X` in the `StatefulPartitionedCall:X` to get the correct order. I don't know if this is the best and safest way to get the original ordering.

**Comment by charleswg** [comment]
For me that doesn’t work. But the index order works in your case too.

**Comment by ydshieh** [comment]
@charleswg ,  index order doesn't work in my case above, because the ascending order of index is the descending order of my keras model output.

I hope TFLite team can let us specify explicitly the input/output orders, for the sake of the ML production stability.

**Comment by Saduf2019** [comment]
@shayantabatabaee 
Could you please try on latest stable version of tf and let us know if this is still an issue.Thanks!

**Comment by tgpsantos** [comment]
> 
> 
> @shayantabatabaee
> Could you please try on latest stable version of tf and let us know if this is still an issue.Thanks!

We've tried it and it looks like it's fixed :) Thanks!

**Comment by shayantabatabaee** [comment]
Yes I tested it ,and it seems to be fixed
Thanks

**Comment by Saduf2019** [comment]
Thank you for your update, glad the issue is resolved.

**Comment by kibeom7565** [comment]
@shayantabatabaee which tensorflow version did you use? (Issue fixed version)

**Comment by zldrobit** [comment]
I would like to chime in. I am using TensorFlow 2.7.0 to export TFLite models, the output tensor order still changes in the conversion.

**Comment by swarmidentity** [comment]
I struggled with this in flutter tflite for Android - after a lot of failed attempts to change the TFLite generation/rename outputs I gave up and just changed the order of my TFLite interpreter.

https://github.com/shaqian/flutter_tflite/pull/235

**Comment by NAM-hj** [comment]
I use Tensorflow 2.8.0. I have the same problem.
With or Without assigning name of each inputs and outputs of my model, TFLite converter change orders

**Comment by chao-camect** [comment]
I have the same problem with 2.7.0
Baffled why such a simple thing could be so difficult..

**Comment by chao-camect** [comment]
I downgraded to tf2.6.1. It works.  So it's a regression.
The bug is only fixed in 2.5 and 2.6.
@zldrobit @swarmidentity @NAM-hj

**Comment by misterBart** [comment]
Several threads exist about this issue. The issue persists now for more than a year. The proposed solution of using signature defs comes again with its own problems. My two work-arounds:
(1) Use tf.compat.v1.lite.TFLiteConverter
(2) Sort output indices
For details see at https://github.com/tensorflow/tensorflow/issues/57043

**Comment by krunalvaghani** [comment]
Any update, how we can solve the issue? with tensorflow==2.8.0
[Example ](https://www.tensorflow.org/lite/models/modify/model_maker/object_detection)from TensorFlow, with custom data, changes order of the output.

**Comment by pmjandreen** [comment]
Signatures seems to be the only way right now: https://www.tensorflow.org/lite/guide/signatures . If one wants to avoid signatures the proper workaround is to ensure consistent naming of input/output tensors. This seems to be what determines the order. You can then easily check the resulting order in for instance Netron. For completely identical tensors (same spatial resolution and number of channels) one can check the order by trial and error. There is of course no guarantee that the order won't change with a new version of Tensorflow or so.

When using Keras one can name tensors after they have been created like this: 
output = tf.keras.layers.Lambda(lambda x: x, name = 'my_output')(output) . It sometimes comes in handy. But it often makes more sense to name them in the layer that created them of course.

**Comment by AlexanderLutsenko** [comment]
Alphabetical naming does not always work, unfortunately. There is a way to put the inputs/outputs back in the original order, but it has to be done manually:
https://github.com/AlexanderLutsenko/tflite_converter_fixer

**Comment by misterBart** [comment]
Thanks. That may come to use to me when `tf.compat.v1.lite.TFLiteConverter` is no longer supported.
I quickly read your github page, "One could think fixing such a nuisance would be an easy task, but the bug persists since at least 2019, so we might as well learn to live with it." I like your sense of reality :)

**Comment by AlexanderLutsenko** [comment]
@misterBart I see, `tf.compat.v1.lite.TFLiteConverter` does indeed work as intended in my toy example. Unfortunately, it also bugs out with many real-life models.

**Comment by ggn1** [comment]
> I managed to get the order back by sorting the index of each output using tflite_interpreter.get_output_details()

Thanks. This worked for me.

**Comment by eric-sawyer-vifive** [comment]
I'm still seeing this issue in both 2.9 and 2.11. Unfortunately I can't use `tf.compat.v1.lite.TFLiteConverter` for my application. I'll try signatures, but it seems they shouldn't be necessary for this to be fixed.

**Comment by louis925** [comment]
A workaround is that you can read the tflite file and modify the output order manually using [tensorflow/lite/tools/flatbuffer_utils.py](https://github.com/tensorflow/tensorflow/blob/33cadc4f901ad6ff9944764c6301d42e49e4b56c/tensorflow/lite/tools/flatbuffer_utils.py)

Simple example is like:
```python
model = flatbuffer_utils.read_model(input_tflite_path)

# assume only 1 subgraph in the model
subgraph = model.subgraphs[0]

# change order of the output
subgraph.outputs = subgraph.outputs[::-1]  # using reverse as an example

flatbuffer_utils.write_model(model, output_tflite_path)
```

Ideally, you can check the signature to see which node is the node you want.
```python
for output in model.signatureDefs[0].outputs:
  print(output.name)
  print(output.tensorIndex)
```
and modify `subgraph.outputs` based on the `output.name` and `output.tensorIndex`.

**Comment by happyTonakai** [comment]
5 years passed and this bug still exists...
